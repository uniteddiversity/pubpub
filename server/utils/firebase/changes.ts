import { Step } from 'prosemirror-transform';
import { Reference } from '@firebase/database-types';
import { uncompressStepJSON } from 'prosemirror-compress-pubpub';

import { editorSchema } from '../firebaseAdmin';

export const getStepsInChangeRange = async (
	branchRef: Reference,
	startIndex: number,
	endIndex: number,
): Promise<Step[]> => {
	const changesSnapshot = await branchRef
		.child('changes')
		.orderByKey()
		.startAt(startIndex.toString())
		.endAt(endIndex.toString())
		.once('value');
	const changes = changesSnapshot.val();
	if (changes) {
		return Object.keys(changes)
			.sort((a, b) => parseInt(a, 10) - parseInt(b, 10))
			.map((key) => changes[key])
			.map((change) =>
				change.s.map((compressedStep) =>
					Step.fromJSON(editorSchema, uncompressStepJSON(compressedStep)),
				),
			)
			.reduce((a, b) => [...a, ...b], []);
	}
	return [];
};
