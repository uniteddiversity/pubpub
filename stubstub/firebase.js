import { Node } from 'prosemirror-model';
import { Transform } from 'prosemirror-transform';

import { createFirebaseChange, getFirebaseDoc } from 'client/components/Editor';
import { Branch } from 'server/models';
import { getBranchRef, editorSchema } from 'server/utils/firebaseAdmin';

export const editPub = async (pubId) => {
	const draftBranch = await Branch.findOne({ where: { pubId: pubId, title: 'draft' } });
	const branchRef = getBranchRef(pubId, draftBranch.id);
	const firebaseDocInfo = await getFirebaseDoc(branchRef, editorSchema);
	let doc = Node.fromJSON(editorSchema, firebaseDocInfo.doc);
	let currentKey = firebaseDocInfo.key;
	let pendingSteps = [];

	const api = {
		transform: (fn) => {
			const tr = new Transform(doc);
			fn(tr, editorSchema);
			doc = tr.doc;
			pendingSteps.push(...tr.steps);
			return api;
		},
		writeChange: async () => {
			const change = createFirebaseChange(pendingSteps, draftBranch.id, 'stubstub-firebase');
			await branchRef.child(`changes/${currentKey + 1}`).set(change);
			++currentKey;
			pendingSteps = [];
		},
		clearChanges: async () => {
			await branchRef.child(`changes`).remove();
			pendingSteps = [];
			currentKey = -1;
		},
		getDoc: () => {
			return doc;
		},
		getKey: () => {
			return currentKey;
		},
	};

	return api;
};
