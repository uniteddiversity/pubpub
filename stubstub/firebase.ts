import { Reference } from '@firebase/database-types';
import { Node } from 'prosemirror-model';
import { Step, Transform } from 'prosemirror-transform';
import uuid from 'uuid/v4';

import { createFirebaseChange, getFirebaseDoc } from 'client/components/Editor';
import { Branch } from 'server/models';
import { getBranchRef, getDatabaseRef, editorSchema } from 'server/utils/firebaseAdmin';

type EditorSchema = typeof editorSchema;
type TransformFn = (t: Transform<EditorSchema>, s: EditorSchema) => void;

export const editFirebaseDraftByRef = async (ref: Reference, { branchId = 'no-branch' } = {}) => {
	const fetchDoc = async () => {
		const { key, doc: docJson } = await getFirebaseDoc(ref, editorSchema);
		const doc = Node.fromJSON(editorSchema, docJson);
		return { key: key, doc: doc };
	};

	let { doc, key: currentKey } = await fetchDoc();
	let pendingSteps: Step[] = [];

	const api = {
		transform: (fn: TransformFn) => {
			const tr = new Transform(doc);
			fn(tr, editorSchema);
			doc = tr.doc;
			pendingSteps.push(...tr.steps);
			return api;
		},
		writeChange: async () => {
			const change = createFirebaseChange(pendingSteps, branchId, 'stubstub-firebase');
			await ref.child(`changes/${currentKey + 1}`).set(change);
			++currentKey;
			pendingSteps = [];
		},
		clearChanges: async () => {
			await ref.child(`changes`).remove();
			const refetch = await fetchDoc();
			doc = refetch.doc;
			currentKey = refetch.key;
			pendingSteps = [];
		},
		getDoc: () => {
			return doc;
		},
		getKey: () => {
			return currentKey;
		},
		getRef: () => {
			return ref;
		},
	};

	return api;
};

export const editFirebaseDraft = (refKey: string = uuid()) => {
	return editFirebaseDraftByRef(getDatabaseRef(refKey)!);
};

export const editPub = async (pubId: string) => {
	const draftBranch = await Branch.findOne({ where: { pubId: pubId, title: 'draft' } });
	const branchRef = getBranchRef(pubId, draftBranch.id)!;
	return editFirebaseDraftByRef(branchRef, { branchId: draftBranch.id });
};
