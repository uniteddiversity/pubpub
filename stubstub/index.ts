export { setup, teardown } from './prepare';
export { login } from './userToAgentMap';
export { stubModule as stub, stubOut, stubFirebaseAdmin } from './stub';
export { modelize } from './modelize/modelize';
export { determinize } from './determinize';
export { editPub, editFirebaseDraft } from './firebase';
