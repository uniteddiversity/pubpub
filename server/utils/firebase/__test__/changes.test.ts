/* global describe, it, expect */
import { editFirebaseDraft } from 'stubstub';

import { getStepsInChangeRange } from '../changes';

describe('getStepsInChangeRange', () => {
	it('returns expected steps', async () => {
		const editor = await editFirebaseDraft();
		editor.transform((tr, schema) => {
			tr.insert(0, schema.text('Hello fellow human'));
			tr.insert(0, schema.text('Here are some steps'));
		});
		await editor.writeChange();
		editor.transform((tr, schema) => {
			tr.insert(10, schema.text('This is another step'));
			tr.insert(5, schema.text('Very cool'));
		});
		await editor.writeChange();
		editor.transform((tr, schema) => {
			tr.replaceRangeWith(0, 10, schema.text('Okay just one more'));
		});
		await editor.writeChange();
		const [zeroToTwo, negativeOneToTwo] = await Promise.all([
			getStepsInChangeRange(editor.getRef(), 0, 2),
			getStepsInChangeRange(editor.getRef(), -1, 2),
		]);
		expect(zeroToTwo).toEqual(negativeOneToTwo);
		expect(zeroToTwo).toMatchSnapshot();
		expect(await getStepsInChangeRange(editor.getRef(), 0, 0)).toMatchSnapshot();
		expect(await getStepsInChangeRange(editor.getRef(), 0, 1)).toMatchSnapshot();
		expect(await getStepsInChangeRange(editor.getRef(), 1, 2)).toMatchSnapshot();
	});
});
