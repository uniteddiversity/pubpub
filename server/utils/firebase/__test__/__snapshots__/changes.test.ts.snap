// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStepsInChangeRange returns expected steps 1`] = `
Array [
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Hello fellow human",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Here are some steps",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
  Object {
    "from": 10,
    "slice": Object {
      "content": Array [
        Object {
          "text": "This is another step",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 10,
  },
  Object {
    "from": 5,
    "slice": Object {
      "content": Array [
        Object {
          "text": "Very cool",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 5,
  },
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Okay just one more",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "to": 10,
  },
]
`;

exports[`getStepsInChangeRange returns expected steps 2`] = `
Array [
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Hello fellow human",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Here are some steps",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
]
`;

exports[`getStepsInChangeRange returns expected steps 3`] = `
Array [
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Hello fellow human",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Here are some steps",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
    },
    "stepType": "replace",
    "to": 0,
  },
  Object {
    "from": 10,
    "slice": Object {
      "content": Array [
        Object {
          "text": "This is another step",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 10,
  },
  Object {
    "from": 5,
    "slice": Object {
      "content": Array [
        Object {
          "text": "Very cool",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 5,
  },
]
`;

exports[`getStepsInChangeRange returns expected steps 4`] = `
Array [
  Object {
    "from": 10,
    "slice": Object {
      "content": Array [
        Object {
          "text": "This is another step",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 10,
  },
  Object {
    "from": 5,
    "slice": Object {
      "content": Array [
        Object {
          "text": "Very cool",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 5,
  },
  Object {
    "from": 0,
    "slice": Object {
      "content": Array [
        Object {
          "attrs": Object {
            "class": null,
            "id": null,
          },
          "content": Array [
            Object {
              "text": "Okay just one more",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "to": 10,
  },
]
`;
